src/ray/raylet/tests/node_manager_test.cc

```diff
@@ -1219,6 +1217,27 @@ INSTANTIATE_TEST_SUITE_P(PinObjectIDsIdempotencyVariations,
                          PinObjectIDsIdempotencyTest,
                          testing::Bool());
 
+TEST_F(NodeManagerTest, TestNoCrashIfGcsPublishesSelfDeadWhenShuttingDown) {
+  // The raylet shouldn't crash if the node starts shutting down (could be from a sigterm
+  // in which case shutting_down_ is set + GCS Unregister RPC is made externally) and then
+  // the GCS publishes that the node is dead.
+
+  gcs::SubscribeCallback<NodeID, rpc::GcsNodeInfo> publish_node_change_callback;
+  EXPECT_CALL(*mock_gcs_client_->mock_node_accessor, AsyncSubscribeToNodeChange(_, _))
+      .WillOnce([&](const gcs::SubscribeCallback<NodeID, rpc::GcsNodeInfo> &subscribe,
+                    const gcs::StatusCallback &done) {
+        publish_node_change_callback = subscribe;
+      });
+  node_manager_->RegisterGcs();
+
+  shutting_down_ = true;
```

Comment:
can we also pair/add a death test where we don't set `shutting_down_` and expect fatal when not shutting down?